pallets/gear-builtin/src/proxy.rs

```diff
@@ -77,20 +77,18 @@ where
 {
     fn handle(
         dispatch: &StoredDispatch,
-        gas_limit: u64,
-    ) -> (Result<Payload, BuiltinActorError>, u64) {
-        let Ok(request) = Request::decode(&mut dispatch.payload_bytes()) else {
-            return (Err(BuiltinActorError::DecodingError), 0);
-        };
+        context: &mut BuiltinContext,
+    ) -> Result<Payload, BuiltinActorError> {
+        let request = Request::decode(&mut dispatch.payload_bytes())
+            .map_err(|_| BuiltinActorError::DecodingError)?;
 
         let origin = dispatch.source();
 
-        match Self::cast(request) {
-            Ok(call) => {
-                let (result, actual_gas) = Pallet::<T>::dispatch_call(origin, call, gas_limit);
-                (result.map(|_| Default::default()), actual_gas)
-            }
-            Err(e) => (Err(e), gas_limit),
-        }
+        let call = Self::cast(request)?;
+        Pallet::<T>::dispatch_call(origin, call, context).map(|_| Default::default())
+    }
+
+    fn max_gas() -> u64 {
+        Default::default()
```

Comment:
Interesting idea. Instead of the complexity parameter (i.e. the payload size) we could let `max_gas` take some small fraction of the payload. Actually, each `max_gas` could interpret it differently depending whether we gauge the gas cost though benchmarking or theoretical considerations.